File: clrs/2/INSERTION-SORT/insert_sort.py
#!/usr/bin/env python
from timing import timing
from randata import getrandata
import time

@timing
def insert_sort(array):
    length = len(array)
    for i in range(1, length):
        key = array[i]
        j = i - 1
        while j >= 0 and array[j] > key:
            array[j+1] = array[j]
            j = j - 1
        array[j+1] = key

@timing
def insert_sort2(array):    
    ' faster '
    length = len(array)
    for i in range(1, length):
        j = i - 1
        while j >= 0 and array[j] < array[i]:
            j -= 1
        array.insert(j+1, array[i])
        array.pop(i+1)




if __name__ == '__main__':
    arr = getrandata(5000)
    start = time.clock()
    arr.sort(reverse=True)              #fastest
    end = time.clock()
    print(arr)
    print("%s: %fs" % (".sort()", end - start))

    insert_sort2(arr)
    print(arr)


    

File: clrs/2/INSERTION-SORT/insert_sort.c
#include <stdio.h>

void insert_sort(int a[], int n)
{
    int i, j, key;

    for (j = 1; j < n; j++) {
        key = a[j];
        i = j - 1;
        while (i >= 0 && a[i] > key) {
            a[i + 1] = a[i];
            i--;
        }
        a[i + 1] = key;
    }
}

void print_array(int a[], int n)
{
    int i;
    for (i = 0; i < n; i++)
        printf("%d ", a[i]);
    printf("\n");
}

int main()
{
    int a[] = {5, 3, 2, 7, 1, 3};

    print_array(a, 6);

    insert_sort(a, 6);
    print_array(a, 6);

    return 0;
}

    




